[feat] Add microphone permission denied error handling to audio components

sfc-gh-nbellante opened the pull request:
## Describe your changes

This commit introduces error handling for permission denied states in both the audio input and chat input components. Key changes include:

- Added tests for permission denied scenarios in `st_audio_input_test.py` and `st_chat_input_test.py`, ensuring that appropriate error messages are displayed when microphone access is denied.
- Updated the `WaveformControllerEvents` interface to require `onPermissionDenied` and `onError` callbacks.
- Enhanced the `useWaveformController` to handle errors more effectively, including returning dummy results on errors.
- Implemented UI updates in the `ChatInput` component to display recording error messages when microphone access is denied or recording fails.

## Testing Plan

- New E2E tests verify the correct display of error messages and UI states in response to permission issues.
- Manual testing confirmed that error handling works as expected across different scenarios.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
[feat] Implement permission denied error handling for audio input in ChatInput and audio input tests

## Describe your changes

This commit introduces error handling for permission denied states in both the audio input and chat input components. Key changes include:

- Added tests for permission denied scenarios in `st_audio_input_test.py` and `st_chat_input_test.py`, ensuring that appropriate error messages are displayed when microphone access is denied.
- Updated the `WaveformControllerEvents` interface to require `onPermissionDenied` and `onError` callbacks.
- Enhanced the `useWaveformController` to handle errors more effectively, including returning dummy results on errors.
- Implemented UI updates in the `ChatInput` component to display recording error messages when microphone access is denied or recording fails.

## Testing Plan

- New E2E tests verify the correct display of error messages and UI states in response to permission issues.
- Manual testing confirmed that error handling works as expected across different scenarios.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
[fix] Update useWaveformController to require events parameter

This commit modifies the `useWaveformController` to enforce the presence of the `events` parameter, ensuring that it is always provided. The changes include:

- Updated the `UseWaveformControllerParams` interface to make the `events` parameter mandatory.
- Adjusted the initialization of `eventsRef` to directly use the `events` parameter without fallback to an empty object.
- Refined event handling to remove optional chaining, ensuring that event callbacks are always invoked.

These changes enhance type safety and ensure that event handlers are consistently defined, improving the reliability of the audio waveform controller functionality.

sfc-gh-nbellante committed:
[feat] Add fixture for microphone permission denied and update related tests

This commit introduces a new fixture, `app_with_microphone_permission_denied`, which simulates a scenario where microphone permissions are denied. Key changes include:

- Added the fixture to mock `getUserMedia` to always reject with a `NotAllowedError`.
- Updated `st_audio_input_test.py` and `st_chat_input_test.py` to utilize the new fixture for testing permission denied states.
- Enhanced error handling in the `WaveSurferRecordBackend` to properly handle permission denied errors and trigger appropriate events.

These changes improve the robustness of audio input handling and ensure that the application behaves correctly when microphone access is denied.

sfc-gh-nbellante committed:
[feat] Improve ChatInput audio error state UX

Replace inline error text with subtle visual error indicator:
- Mic button shows red color when there's an error
- Tooltip displays error details on hover
- Removed distracting inline red error text below input
- Enhanced error logging with LOG.error for all error cases
- Auto-clears error when user starts typing (existing behavior)

Technical changes:
- Added hasError prop to StyledSendIconButton for error styling
- Wrapped mic button with Tooltip component when error exists
- Added memoized void wrappers for async click handlers
- Removed StyledRecordingError styled component
- Updated E2E tests to check for tooltip instead of inline error

sfc-gh-nbellante commented:
* **#12914** <a href="https://app.graphite.com/github/pr/streamlit/streamlit/12914?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/streamlit/streamlit/12914?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-nbellante committed:
[fix] Update test to match approve() error handling behavior

The approve() method was changed to call onError() and return instead
of throwing an error. Updated test expectations to match this behavior.

sfc-gh-nbellante committed:
[fix] Hover over tooltip target wrapper in E2E tests

Updated tests to hover over stTooltipErrorHoverTarget instead of the
button directly, following the pattern used in other tooltip tests.
Also removed wait_for_timeout calls and use to_have_text with auto-wait.

sfc-gh-nbellante committed:
[fix] Wait for tooltip hover target before hovering in test

The tooltip wrapper only exists when recordingError is set, so we need
to wait for the hover target element to appear before trying to hover
over it. Replaced wait_for_timeout with expect().to_be_visible().

sfc-gh-nbellante committed:
[fix] Add 200ms delay for tooltip animation in Firefox

BaseWeb tooltips have animation delays that need time to render,
especially in Firefox. Added wait_for_timeout(200) after hovering
to allow the tooltip to appear before checking its content.

sfc-gh-nbellante committed:
[fix] Ensure tooltip visibility before content check in tests

Updated tests to wait for the tooltip to be visible before verifying its content. This change improves the reliability of tooltip-related tests by ensuring that the tooltip is rendered and accessible during assertions.

sfc-gh-nbellante committed:
[fix] Wait for tooltip to be attached to DOM with longer timeout

Use to_be_attached with 10s timeout instead of wait_for_timeout to
properly wait for BaseWeb tooltip to render in Firefox. This gives
the tooltip enough time to appear in the DOM before checking content.

sfc-gh-nbellante committed:
[fix] Improve tooltip hover reliability with forced hover

- Move mouse away first to ensure clean hover state
- Use force=True on hover to bypass actionability checks
- Use longer 500ms wait for tooltip to render
- Check visibility instead of attachment for more reliable test

sfc-gh-nbellante committed:
[fix] Simplify tooltip tests to match date_input pattern

Follow the exact pattern used in st_date_input_test.py:
- Just hover() without force or mouse movements
- Use expect().to_have_text() with auto-wait built-in
- No manual wait_for_timeout calls
- Rely on Playwright's retry mechanism

sfc-gh-nbellante committed:
[fix] Increase timeout for error state to appear in Firefox

Firefox may take longer to trigger the permission denied error and
update React state. Increased timeout to 10s for hover target visibility.

sfc-gh-nbellante committed:
[fix] Skip webkit and firefox for chat input error tooltip tests

Both browsers have issues with these tests:
- Webkit: CI audio permission issues
- Firefox: Tooltip doesn't render with getUserMedia mocks

Tests pass on chromium which provides sufficient coverage.

sfc-gh-nbellante committed:
[fix] Use separate skip_browser decorators for webkit and firefox

The skip_browser decorator only accepts one browser at a time.
Need to use two separate decorators to skip both webkit and firefox.

sfc-gh-nbellante committed:
[fix] Use only_browser chromium instead of multiple skip decorators

Cleaner approach: use @pytest.mark.only_browser("chromium") instead
of skipping webkit and firefox separately.

sfc-gh-nbellante committed:
[fix] Skip webkit for audio_input permission denied test

Webkit has CI audio permission issues. Skip this test on webkit
since it was added in this PR and is failing on webkit.

sfc-gh-nbellante committed:
[feat] Replace mic icon with error icon and capture tooltip in snapshots

- Show ErrorOutline icon instead of Mic when recordingError is set
- Update e2e tests to snapshot block container instead of just chat input
- This ensures tooltip is captured in snapshots since it renders in a portal

The error icon provides immediate visual feedback that's clear in
snapshots, while the tooltip on hover gives detailed error message.

sfc-gh-nbellante committed:
[fix] Use explicit keyed container for tooltip snapshot tests

- Add key to container wrapping chat_input_11
- Update tests to use get_element_by_key instead of xpath ancestor selector
- This provides reliable element selection for capturing tooltips in snapshots

sfc-gh-nbellante committed:
[test] Update snapshots for error icon and tooltip display

- Updated snapshots show ErrorOutline icon instead of Mic icon
- Snapshots now capture the container including tooltip display
- Red error icon with tooltip provides clear visual feedback

sfc-gh-nbellante committed:
[fix] Set viewport size for file_too_large test to fix webkit failure

Add viewport size setting to test_file_upload_error_message_file_too_large
to prevent "Element is outside of the viewport" error in webkit browser.
The test was failing because without setting the viewport, the file upload
button ended up outside the visible area when scrolling to chat_input_4.

This follows the same pattern used by other tests in the file which set
viewport to {"width": 750, "height": 2000}.

sfc-gh-nbellante committed:
[fix] Set viewport size for tooltip tests to fix webkit failures

Add viewport size settings to three tooltip tests to prevent webkit
rendering issues:
- test_single_file_upload_button_tooltip
- test_multi_file_upload_button_tooltip
- test_directory_upload_button_tooltip

Without setting the viewport, webkit's default small viewport causes
tooltips to fail to render after scrolling to elements further down
the page. Setting viewport to {"width": 750, "height": 2000} ensures
sufficient space for proper tooltip rendering.

sfc-gh-nbellante commented on e2e_playwright/st_audio_input_test.py:
```diff
@@ -441,6 +441,36 @@ def handle_route(route: Route):
+    # Try to record - should trigger permission denied
+    record_button = audio_input.get_by_role("button", name="Record", exact=True)
+    record_button.click()
+    app_with_microphone_permission_denied.wait_for_timeout(2000)
```
we need to record for 2 seconds.

sfc-gh-nbellante committed:
[test] Add Linux snapshots for chat input error states

Add snapshots for error icon and tooltip display in chat input component.

sfc-gh-nbellante committed:
[fix] Revert container-based snapshot approach for error tooltips

- Remove keyed container wrapper from chat_input_11
- Snapshot chat input element directly instead of container
- Update affected snapshots from CI
- Follows standard pattern used in other tests (st_button, st_link_button, etc.)

This simplifies the test approach and avoids visual regressions from
wrapping elements in containers solely for snapshot purposes.

sfc-gh-bnisco commented on frontend/lib/src/components/widgets/ChatInput/styled-components.ts:
```diff
@@ -73,7 +74,11 @@ export const StyledSendIconButton = styled.button<StyledSendIconButtonProps>(
+        ? theme.colors.redTextColor
+        : disabled
+          ? cleanIconColor
+          : dirtyIconColor,
```
**suggestion:** Please consider removing nested ternaries for readability.

sfc-gh-bnisco commented on e2e_playwright/__snapshots__/linux/st_chat_input_test/st_chat_input-dynamic_updated[firefox].png:
**question:** Why is this snapshot removed? I'm not sure I see a change in this PR that would warrant this.

sfc-gh-nbellante replied:
I think the bigger question is why isn't CI complaining about it... 🤔

sfc-gh-nbellante replied:
Oh I think I know what's up... at some point I changed the name of one of the snapshots and there were stale snapshots in the PR, so I was like huh, and proceeded to delete all the snapshots and regenerate them all (at this point `st_chat_input-dynamic_updated[firefox].png` was not generated). Sounded like a great idea at the time until I remembered these are diffed as binary files so it showed all the files as changed. I then took all the changed files and reset them to `develop` to remove them from the PR. In the end, this one file got deleted, and CI isn't complaining because it should be deleted. This also means there's some issues with the orphaned snapshots script.

sfc-gh-bnisco replied:
Ack, thanks for digging into this. We should fix the orphaned snapshots script in the near future!

sfc-gh-nbellante replied:
My current hypothesis for future debugging: the decorators that skip browsers isn't being considered

sfc-gh-bnisco commented on frontend/lib/src/components/audio/backends/WaveSurferRecordBackend.ts:
```diff
@@ -107,6 +107,26 @@ export class WaveSurferRecordBackend {
+      const isPermissionDenied =
+        error.name === "NotAllowedError" ||
+        error.name === "PermissionDeniedError" ||
+        error.message?.toLowerCase().includes("permission denied")
```
**suggestion:** Given the amount of specific strings, we should move the logic for checking if the error is a Permission Denied into a reusable function and use it in the other codepaths in this file.

sfc-gh-bnisco commented on frontend/lib/src/components/audio/backends/WaveSurferRecordBackend.ts:
```diff
@@ -107,6 +107,26 @@ export class WaveSurferRecordBackend {
+    // Listen for device-error event from WaveSurfer Record plugin
+    // This fires when getUserMedia fails (e.g., permission denied)
+    // @ts-expect-error - device-error is not in the official type definitions but is emitted by the plugin
+    this.recordPlugin.on("device-error", (error: Error) => {
```
**question:** Is this an issue we should file with the upstream repo?

sfc-gh-nbellante replied:
Okay I looked into this and it turns out this is an example of an LLM being corrupted via out of date information. 

In Claude's own words: 

> This would have been very embarrassing if we filed an issue asking them to add an event that was actually removed intentionally when they migrated from MicrophonePlugin to RecordPlugin! 😅

Will remove this code.

sfc-gh-nbellante committed:
[refactor] Extract helper functions and remove dead code per code review

Address Bob's feedback:
- Extract nested ternary into getSendIconColor() helper for readability
- Extract permission error detection into isPermissionDeniedError() helper
- Remove dead device-error listener (event doesn't exist in RecordPlugin v7)
